In [1]:
import pandas as pd
import numpy as np
import plotly.io as pio
import matplotlib.pyplot as plt
import re
from sklearn.impute import SimpleImputer
pio.renderers.default = 'notebook'

In [2]:
df_links = pd.read_csv('datasets/links_first_page.csv')
df_data = pd.read_csv('datasets/cars_df.csv')

In [3]:
df_links.head()

title        price  \
0  Kia K900 3.3 AT, 2019, 91 037 км, с пробегом, ...  2 999 000 ₽   
1  Mercedes-Benz V-класс 2.1 AT, 2017, 95 250 км,...  5 650 000 ₽   
2  Opel Astra GTC 1.4 AT, 2013, 113 000 км, с про...    699 000 ₽   
3  Volkswagen Tiguan 2.0 AT, 2015, 165 000 км, с ...  1 195 000 ₽   
4  Ford Kuga 1.5 AT, 2018, 103 000 км, с пробегом...  1 780 000 ₽   

                                                link  
0  https://www.avito.ru/moskva/avtomobili/kia_k90...  
1  https://www.avito.ru/moskva/avtomobili/mercede...  
2  https://www.avito.ru/moskva/avtomobili/opel_as...  
3  https://www.avito.ru/moskva/avtomobili/volkswa...  
4  https://www.avito.ru/moskva/avtomobili/ford_ku...

In [4]:

df_data.head()

title  Год выпуска  \
0  Kia K900 3.3 AT, 2019, 91 037 км, с пробегом, ...       2019.0   
1  Mercedes-Benz V-класс 2.1 AT, 2017, 95 250 км,...       2017.0   
2  Opel Astra GTC 1.4 AT, 2013, 113 000 км, с про...       2013.0   
3  Volkswagen Tiguan 2.0 AT, 2015, 165 000 км, с ...       2015.0   
4  Ford Kuga 1.5 AT, 2018, 103 000 км, с пробегом...       2018.0   

                   Поколение      Пробег               История пробега  \
0             II (2018—2022)   91 037 км   8 записей в отчёте Автотеки   
1           W447 (2014—2024)   95 250 км  19 записей в отчёте Автотеки   
2   J рестайлинг (2011—2017)  113 000 км   6 записей в отчёте Автотеки   
3   I рестайлинг (2011—2019)  165 000 км   9 записей в отчёте Автотеки   
4  II рестайлинг (2016—2019)  103 000 км  16 записей в отчёте Автотеки   

  Владельцев по ПТС Состояние                          Модификация  \
0                 2  Не битый                3.3 4WD AT (249 л.с.)   
1                 2  Не битый  V 250 d 2.1 4MATIC AT L3 (190 л.с.)   
2                 2  Не битый              1.4 Turbo AT (140 л.с.)   
3                 2  Не битый        2.0 TDI 4Motion AT (140 л.с.)   
4                 1  Не битый       1.5 EcoBoost 4WD AT (150 л.с.)   

  Объём двигателя Тип двигателя  ... VIN или номер кузова          Обмен  \
0           3.3 л        Бензин  ...    XWES*************       Возможен   
1           2.1 л        Дизель  ...    WDF4*************  Не интересует   
2           1.4 л        Бензин  ...    XWFP*************       Возможен   
3             2 л        Дизель  ...    XW8Z*************       Возможен   
4           1.5 л        Бензин  ...    Z6FA*************              0   

                                     Оценка          пробег_история  \
0  2 999 000 ₽ — соответствует оценке Авито   8 записей с 2020 года   
1  5 650 000 ₽ — соответствует оценке Авито  19 записей с 2019 года   
2        699 000 ₽ — цена ниже Авито Оценки   6 записей с 2017 года   
3      1 195 000 ₽ — цена ниже Авито Оценки   9 записей с 2016 года   
4  1 780 000 ₽ — соответствует оценке Авито  16 записей с 2018 года   

              ДТП      другое_история                   тех_история  \
0  ДТП не найдены  2 владельца по ПТС    4 записи с описанием работ   
1             NaN                 NaN  19 записей с описанием работ   
2  ДТП не найдены                 NaN                           NaN   
3             NaN  2 владельца по ПТС   5 записей с описанием работ   
4             NaN   1 владелец по ПТС  10 записей с описанием работ   

                         sold  \
0   1 раз продавался на Авито   
1                         NaN   
2  3 раза продавался на Авито   
3   1 раз продавался на Авито   
4   1 раз продавался на Авито   

                                               Метро       ПТС  
0  Аннино:6–10 мин.;Лесопарковая:11–15 мин.;Улица...       NaN  
1  Проспект Вернадского:до 5 мин.;Новаторская:16–...  Оригинал  
2  Южная:11–15 мин.;Чертановская:21–30 мин.;Пражс...       NaN  
3  Волоколамская:16–20 мин.;Мякинино:от 31 мин.;Т...       NaN  
4  Ховрино:до 5 мин.;Беломорская:21–30 мин.;Речно...  Оригинал  

[5 rows x 26 columns]

In [5]:
df = pd.merge(df_data, df_links, on='title', how='inner')
df.head()

title  Год выпуска  \
0  Kia K900 3.3 AT, 2019, 91 037 км, с пробегом, ...       2019.0   
1  Mercedes-Benz V-класс 2.1 AT, 2017, 95 250 км,...       2017.0   
2  Opel Astra GTC 1.4 AT, 2013, 113 000 км, с про...       2013.0   
3  Volkswagen Tiguan 2.0 AT, 2015, 165 000 км, с ...       2015.0   
4  Ford Kuga 1.5 AT, 2018, 103 000 км, с пробегом...       2018.0   

                   Поколение      Пробег               История пробега  \
0             II (2018—2022)   91 037 км   8 записей в отчёте Автотеки   
1           W447 (2014—2024)   95 250 км  19 записей в отчёте Автотеки   
2   J рестайлинг (2011—2017)  113 000 км   6 записей в отчёте Автотеки   
3   I рестайлинг (2011—2019)  165 000 км   9 записей в отчёте Автотеки   
4  II рестайлинг (2016—2019)  103 000 км  16 записей в отчёте Автотеки   

  Владельцев по ПТС Состояние                          Модификация  \
0                 2  Не битый                3.3 4WD AT (249 л.с.)   
1                 2  Не битый  V 250 d 2.1 4MATIC AT L3 (190 л.с.)   
2                 2  Не битый              1.4 Turbo AT (140 л.с.)   
3                 2  Не битый        2.0 TDI 4Motion AT (140 л.с.)   
4                 1  Не битый       1.5 EcoBoost 4WD AT (150 л.с.)   

  Объём двигателя Тип двигателя  ...  \
0           3.3 л        Бензин  ...   
1           2.1 л        Дизель  ...   
2           1.4 л        Бензин  ...   
3             2 л        Дизель  ...   
4           1.5 л        Бензин  ...   

                                     Оценка          пробег_история  \
0  2 999 000 ₽ — соответствует оценке Авито   8 записей с 2020 года   
1  5 650 000 ₽ — соответствует оценке Авито  19 записей с 2019 года   
2        699 000 ₽ — цена ниже Авито Оценки   6 записей с 2017 года   
3      1 195 000 ₽ — цена ниже Авито Оценки   9 записей с 2016 года   
4  1 780 000 ₽ — соответствует оценке Авито  16 записей с 2018 года   

              ДТП      другое_история                   тех_история  \
0  ДТП не найдены  2 владельца по ПТС    4 записи с описанием работ   
1             NaN                 NaN  19 записей с описанием работ   
2  ДТП не найдены                 NaN                           NaN   
3             NaN  2 владельца по ПТС   5 записей с описанием работ   
4             NaN   1 владелец по ПТС  10 записей с описанием работ   

                         sold  \
0   1 раз продавался на Авито   
1                         NaN   
2  3 раза продавался на Авито   
3   1 раз продавался на Авито   
4   1 раз продавался на Авито   

                                               Метро       ПТС        price  \
0  Аннино:6–10 мин.;Лесопарковая:11–15 мин.;Улица...       NaN  2 999 000 ₽   
1  Проспект Вернадского:до 5 мин.;Новаторская:16–...  Оригинал  5 650 000 ₽   
2  Южная:11–15 мин.;Чертановская:21–30 мин.;Пражс...       NaN    699 000 ₽   
3  Волоколамская:16–20 мин.;Мякинино:от 31 мин.;Т...       NaN  1 195 000 ₽   
4  Ховрино:до 5 мин.;Беломорская:21–30 мин.;Речно...  Оригинал  1 780 000 ₽   

                                                link  
0  https://www.avito.ru/moskva/avtomobili/kia_k90...  
1  https://www.avito.ru/moskva/avtomobili/mercede...  
2  https://www.avito.ru/moskva/avtomobili/opel_as...  
3  https://www.avito.ru/moskva/avtomobili/volkswa...  
4  https://www.avito.ru/moskva/avtomobili/ford_ku...  

[5 rows x 28 columns]

проверяем что колонки не потерялись:

In [6]:
df.shape

(1805, 28)

Посомтрим на общую информацию о датасете:

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1805 entries, 0 to 1804
Data columns (total 28 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   title                 1805 non-null   object 
 1   Год выпуска           1751 non-null   float64
 2   Поколение             1749 non-null   object 
 3   Пробег                1748 non-null   object 
 4   История пробега       1675 non-null   object 
 5   Владельцев по ПТС     1734 non-null   object 
 6   Состояние             1714 non-null   object 
 7   Модификация           1707 non-null   object 
 8   Объём двигателя       1687 non-null   object 
 9   Тип двигателя         1699 non-null   object 
 10  Коробка передач       1699 non-null   object 
 11  Привод                1699 non-null   object 
 12  Комплектация          1296 non-null   object 
 13  Тип кузова            1699 non-null   object 
 14  Цвет                  1699 non-null   object 
 15  Руль                 

### Предобработка датасета
Видим, что почти все колонки имеют тип данных объект (из-за того что мы не приводили тип данных при парсинге, что мы исправим чуть позже).

Изначально лучше сразу выбросить те строки, в которых нет никаких данных, кроме информации с главной страницы Авито (то есть те, в которых не None параметрами являются title, price и link), а также параметрами тип кузова, коробка передач, привод. Также можем сразу дропнуть столбец link, так как вдальнейшем он нам не пригодится

In [8]:

bad_rows = df[df[['Тип кузова', 'Коробка передач', 'Привод']].isnull().all(axis=1)]
bad_idx = bad_rows.index

In [9]:
len(bad_idx) # длина равна 54, что соответствует разнице между длиной всего датасета и 

106

In [10]:
bad_idx

Index([  59,   82,  108,  121,  150,  172,  189,  192,  193,  209,
       ...
       1729, 1738, 1748, 1751, 1757, 1767, 1781, 1791, 1801, 1804],
      dtype='int64', length=106)

In [11]:
df.drop(bad_idx,axis=0, inplace=True)
df.drop('link', axis=1, inplace=True)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1699 entries, 0 to 1803
Data columns (total 27 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   title                 1699 non-null   object 
 1   Год выпуска           1699 non-null   float64
 2   Поколение             1699 non-null   object 
 3   Пробег                1698 non-null   object 
 4   История пробега       1632 non-null   object 
 5   Владельцев по ПТС     1698 non-null   object 
 6   Состояние             1699 non-null   object 
 7   Модификация           1699 non-null   object 
 8   Объём двигателя       1687 non-null   object 
 9   Тип двигателя         1699 non-null   object 
 10  Коробка передач       1699 non-null   object 
 11  Привод                1699 non-null   object 
 12  Комплектация          1296 non-null   object 
 13  Тип кузова            1699 non-null   object 
 14  Цвет                  1699 non-null   object 
 15  Руль                  1699

### Обработка значений


Сейчас все данные являются объектами из-за формата парсинга. Хочется четко отделить категориальные и числовые признаки, а для этого надо изменить значения, которые сейчас хранятся в колонках.

#### title
Начнем с title, так как он хранит информацию не только о модели, но и о пробеге и просто часть, связанную с объявлением

In [13]:
df.loc[0].title # видим странные символы => надо их заменить

'Kia K900 3.3 AT, 2019, 91\xa0037\xa0км, с пробегом, цена 2\xa0999\xa0000 руб. - Автомобили в Москве'

In [14]:
df.loc[0].title.replace('\xa0', '') # замена

'Kia K900 3.3 AT, 2019, 91037км, с пробегом, цена 2999000 руб. - Автомобили в Москве'

То есть информация хранится в формате: Модель, год, пробег, наличие пробега, цена - дополнение при парсине (Автомобили в Москве). Пробег хранится в отдельном признаке "Пробег", однако он отсутствует у 3х строчек. Посмотрим на них

In [15]:
df[df['Пробег'].isna()] # видим, что данные строки имеют пропуски по псех своих значениях, поэтому их удаление также релевантно


title  Год выпуска  \
96  Новый Kia Seltos 1.6 AT, 2022, цена от 1 884 9...       2022.0   

        Поколение Пробег История пробега Владельцев по ПТС Состояние  \
96  I (2019—2022)    NaN             NaN               NaN  Не битый   

          Модификация Объём двигателя Тип двигателя  ...     Обмен Оценка  \
96  1.6 AT (123 л.с.)           1.6 л        Бензин  ...  Возможен    NaN   

   пробег_история  ДТП другое_история тех_история sold  \
96            NaN  NaN            NaN         NaN  NaN   

                                                Метро  ПТС           price  
96  Угрешская:6–10 мин.;Волгоградский проспект:6–1...  NaN  от 1 884 900 ₽  

[1 rows x 27 columns]

In [16]:
df.drop(df[df['Пробег'].isna()].index, axis=0, inplace=True)

Теперь можем просто оставить в качестве title весь текст до года выпуска (то есть только название машины)

In [17]:
df['title'] = df['title'].apply(lambda row: row.split(', ')[0])
df.head(1)

title  Год выпуска       Поколение     Пробег  \
0  Kia K900 3.3 AT       2019.0  II (2018—2022)  91 037 км   

               История пробега Владельцев по ПТС Состояние  \
0  8 записей в отчёте Автотеки                 2  Не битый   

             Модификация Объём двигателя Тип двигателя  ...     Обмен  \
0  3.3 4WD AT (249 л.с.)           3.3 л        Бензин  ...  Возможен   

                                     Оценка         пробег_история  \
0  2 999 000 ₽ — соответствует оценке Авито  8 записей с 2020 года   

              ДТП      другое_история                 тех_история  \
0  ДТП не найдены  2 владельца по ПТС  4 записи с описанием работ   

                        sold  \
0  1 раз продавался на Авито   

                                               Метро  ПТС        price  
0  Аннино:6–10 мин.;Лесопарковая:11–15 мин.;Улица...  NaN  2 999 000 ₽  

[1 rows x 27 columns]

#### Год выпуска

Данное значение просто приведем к целочисленному типу, так как будет более релевантно его использовать и переименуем поле в year

In [18]:
df['Год выпуска'] = df['Год выпуска'].astype(int)
df.head(1)

title  Год выпуска       Поколение     Пробег  \
0  Kia K900 3.3 AT         2019  II (2018—2022)  91 037 км   

               История пробега Владельцев по ПТС Состояние  \
0  8 записей в отчёте Автотеки                 2  Не битый   

             Модификация Объём двигателя Тип двигателя  ...     Обмен  \
0  3.3 4WD AT (249 л.с.)           3.3 л        Бензин  ...  Возможен   

                                     Оценка         пробег_история  \
0  2 999 000 ₽ — соответствует оценке Авито  8 записей с 2020 года   

              ДТП      другое_история                 тех_история  \
0  ДТП не найдены  2 владельца по ПТС  4 записи с описанием работ   

                        sold  \
0  1 раз продавался на Авито   

                                               Метро  ПТС        price  
0  Аннино:6–10 мин.;Лесопарковая:11–15 мин.;Улица...  NaN  2 999 000 ₽  

[1 rows x 27 columns]

In [19]:
df.rename(columns={'Год выпуска':'year'}, inplace=True)

#### Поколение

In [20]:
df['Поколение'].value_counts(dropna=False)

Поколение
III (2010—2014)                29
IV рестайлинг (2018—2020)      24
III (2011—2015)                20
II (2017—2020)                 18
V рестайлинг (2015—2020)       18
                               ..
XV30 (2001—2006)                1
I (2017—2023)                   1
W221 рестайлинг (2009—2013)     1
III рестайлинг (2005—2009)      1
VIII рестайлинг (2011—2013)     1
Name: count, Length: 564, dtype: int64

Поколение можно отнести как к категориальной, так и числовой переменной (как номер). Первоначально давайте попробуем выделить из него номер и название поколения (без годов)

In [21]:
def get_age(row):
    if type(row) is str:
        return row[:row.find(' (')]
    return None

df['Поколение'] = df['Поколение'].apply(get_age)

посмотрим на пустые значения

In [22]:
df['Поколение'].isna().sum()

0

заменим пустые значения на наиболее частое (так как процент пустых значений очень мал – 3%)

In [23]:
df['Поколение'].fillna(df['Поколение'].value_counts().index[0], inplace=True)
df.rename(columns={'Поколение':'age'}, inplace=True)

#### Пробег

In [24]:
df['Пробег'].values

array(['91 037 км', '95 250 км', '113 000 км', ..., '159 500 км',
       '122 463 км', '157 500 км'], dtype=object)

Видим, что числа являются числами с пробелами и измерениями км, то есть надо оставить только число


In [25]:
def get_run(row):
    if type(row) is str:
        row = row.rstrip(' км')
        return int(''.join(row.strip().split()))
    return None

df['Пробег'] = df['Пробег'].apply(get_run)

Заполним None значения средним:

In [26]:
df['Пробег'].isna().sum()

0

In [27]:
mean_run = df['Пробег'].mean()

df['Пробег'].fillna(mean_run, inplace=True)
df.rename(columns={'Пробег':'run'}, inplace=True)

#### История пробега / пробег_история

Сначала посмотрим на эти 2 столбца:

In [28]:
df['История пробега'].values

array(['8 записей в отчёте Автотеки', '19 записей в отчёте Автотеки',
       '6 записей в отчёте Автотеки', ..., '20 записей в отчёте Автотеки',
       '6 записей в отчёте Автотеки', '11 записей в отчёте Автотеки'],
      dtype=object)

In [29]:
df['пробег_история'].values

array(['8 записей с 2020 года', '19 записей с 2019 года',
       '6 записей с 2017 года', ..., '20 записей с 2013 года',
       '6 записей с 2016 года', '11 записей с 2017 года'], dtype=object)

видим что записи практически одинаковые и отличаются наличием или отсутствием первой истории о пробеге, однако лучше оставить число записей в колонке История пробега

In [30]:
df.drop(columns=['пробег_история'], inplace=True)
df.rename(columns={'История пробега':'history'}, inplace=True)

In [31]:
df['history'].isna().sum()

66

возьмем только число из истории и заполним None средними

In [32]:
def get_history(row):
    if type(row) is str and 'Проверить' not in row:
        return int(row.strip().split()[0])
    return None

df['history'] = df['history'].apply(get_history)
mean_history = df['history'].mean()
df['history'].fillna(mean_history, inplace=True)

#### Число владельцев по ПТС

In [33]:
df['Владельцев по ПТС'].value_counts(dropna=False)

Владельцев по ПТС
1     723
2     486
3     250
4+    239
Name: count, dtype: int64

категориальная переменная, всего 14 пустых значений (=> можем заполнить самым частым)

In [34]:
df['Владельцев по ПТС'].fillna(df['Владельцев по ПТС'].value_counts().index[0], inplace=True)
df.rename(columns={'Владельцев по ПТС':'owners_pts'}, inplace=True)

#### тип ПТС

In [35]:
df['ПТС'].value_counts(dropna=False)

ПТС
NaN            740
Оригинал       727
Дубликат       118
Электронный    113
Name: count, dtype: int64

В данной ситуации Nan будет релевантнее объединить в другую категорию (так как число Nan значений больше чем значений в другой большой группе, но удаление данного признака может быть нерелевантно)

In [36]:
df['ПТС'].fillna('Другое', inplace=True)
df.rename(columns={'ПТС':'pts_type'}, inplace=True)

#### Состояние

In [37]:
df['Состояние'].value_counts(dropna=False)

Состояние
Не битый    1686
Битый         12
Name: count, dtype: int64

В данном признаке очень много значений одного типа, при этом вариантов "Битый" достаточно мало, поэтому удалим его

In [38]:
df.drop(columns=['Состояние'], inplace=True)

#### Модификация

In [39]:
df['Модификация'].value_counts(dropna=False)

Модификация
2.0 4WD AT (150 л.с.)                      67
2.0 AT (150 л.с.)                          58
1.6 AT (123 л.с.)                          57
1.6 MPI AT (110 л.с.)                      40
2.0 MPI AT (150 л.с.)                      17
                                           ..
1.5 dCi 4WD MT (109 л.с.)                   1
45 TDI 3.0 quattro Tiptronic (249 л.с.)     1
C 220 CDI 2.2 AT (150 л.с.)                 1
1.4 TSI 4WD DSG (150 л.с.)                  1
2.0 4WD AT (226 л.с.)                       1
Name: count, Length: 740, dtype: int64

вынесем ыисло лошадиных сил отдельно, а саму модификацию оставим без изменений. воспльзуемся регулярными выражениями, чтобы выделить число лошадиных сил.

In [40]:
df['horse_power'] = df['Модификация'].apply(lambda x: int(re.search(r'\((\d+)\s*л\.с\.\)', x).group(1)) if re.search(r'\((\d+)\s*л\.с\.\)', x) else None)

df['Модификация'] = df['Модификация'].apply(lambda x: re.sub(r'\s*\(\d+\s*л\.с\.\)', '', x))
df.rename(columns={'Модификация': 'modification'}, inplace=True)

In [41]:

mean_horse = df['horse_power'].mean()

df['horse_power'].fillna(mean_horse, inplace=True)
df['horse_power'] = df['horse_power'].astype(int)


In [42]:
df.rename(columns={'Модификация': 'modification'}, inplace=True)

#### Объем двигателя

видим что объем двигателя в формате float с пробелом и л, переведем просто в формат float:

In [43]:
def get_volume(row):
    if type(row) is str:
        return float(row.strip().split()[0])
    return None

df['Объём двигателя'] = df['Объём двигателя'].apply(get_volume)


посмотрим на число нулевых значений

In [44]:
df['Объём двигателя'].isna().sum()

12

заполним их средним (и переименуем колонку)

In [45]:
mean_volume = df['Объём двигателя'].mean()

df['Объём двигателя'].fillna(mean_volume, inplace=True)
df.rename(columns={'Объём двигателя':'volume'}, inplace=True)

#### Тип двигателя

In [46]:
df['Тип двигателя'].value_counts(dropna=False)

Тип двигателя
Бензин     1394
Дизель      292
Гибрид        9
Газ           2
Электро       1
Name: count, dtype: int64

Можем объединить категории Газ / Гибрид и Электро в "другое"

In [47]:
df['Тип двигателя'] = df['Тип двигателя'].map({'Гибрид': 'другое', 'Газ': 'другое', 'Электро':'другое',
                                               'Бензин': 'Бензин', 'Дизель':'Дизель'})
df['Тип двигателя'].fillna('другое', inplace=True)

df.rename(columns={'Тип двигателя':'engine_type'}, inplace=True)

#### Коробка передач

In [48]:
df['Коробка передач'].value_counts(dropna=False)

Коробка передач
Автомат     1115
Механика     294
Робот        170
Вариатор     119
Name: count, dtype: int64

В данном признаке объединим None в другую категорию, переименуем его только на transmission (и также сразу переименуем все следующие колонки)

In [49]:
df['Коробка передач'].fillna('другое', inplace=True)

In [50]:
df.rename(columns={'Коробка передач': 'transmission', 'Привод':'gear',
                   'Комплектация': 'complete_set', 'Тип кузова': 'body', 
                   'Цвет': 'color', 'Руль': 'wheel', 'VIN или номер кузова': 'number',
                   'Обмен':'exchange', 'Оценка':'mark', 'ДТП':'accident', 'другое_история':'other_history',
                   'тех_история':'tech_history', 'Метро':'subway'}, inplace=True)


#### Привод

In [51]:
df['gear'].value_counts(dropna=False)

gear
Передний    940
Полный      649
Задний      109
Name: count, dtype: int64

пустые значения объединяем в категорию 'другое'

In [52]:
df['gear'].fillna('другое', inplace=True)

#### Комплектация

In [53]:
df['complete_set'].value_counts(dropna=False)

complete_set
NaN                            403
Базовая                        344
Comfort                         57
Base                            56
Prestige                        36
                              ... 
Люкс 21703-23-046                1
Comfort Cruise                   1
Blueefficiency Особая серия      1
SE+ Sport                        1
Status                           1
Name: count, Length: 263, dtype: int64

Очень большое число мелких значений, а число пропущенных значений больше самой крупной категории => можем удалить колонку

In [54]:
df.isna().sum()

title              0
year               0
age                0
run                0
history            0
owners_pts         0
modification       0
volume             0
engine_type        0
transmission       0
gear               0
complete_set     403
body               0
color              0
wheel              0
number             0
exchange           0
mark             209
accident         788
other_history    525
tech_history     644
sold             596
subway            23
pts_type           0
price              0
horse_power        0
dtype: int64

In [55]:
df.drop(columns=['complete_set'], inplace=True)

#### Тип кузова

In [56]:
df['body'].value_counts(dropna=False)

body
Внедорожник 5-дверный    648
Седан                    602
Хетчбек 5-дверный        151
Универсал                 73
Лифтбек                   71
Минивэн                   68
Фургон                    31
Хетчбек 3-дверный         24
Купе                      20
Пикап                      5
Кабриолет                  2
Внедорожник 3-дверный      2
Микроавтобус               1
Name: count, dtype: int64

объединим часть категорий

In [57]:
# для объединения использовался chat gpt, так как я не разбираюсь в машинах
body_mapping = {
    'Внедорожник 5-дверный': 'Внедорожник',
    'Внедорожник 3-дверный': 'Внедорожник',
    'Хетчбек 5-дверный': 'Хэтчбек',
    'Хетчбек 3-дверный': 'Хэтчбек',
    'Лифтбек': 'Хэтчбек',
    'Минивэн': 'Minivan',
    'Фургон': 'Van',
    'Микроавтобус': 'Van'
}

# Применим объединение категорий к колонке 'body'
df['body'] = df['body'].map(body_mapping).fillna(df['body'])


#### Цвет

In [58]:
df['color'].value_counts(dropna=False)

color
Чёрный        536
Белый         382
Серый         233
Серебряный    136
Синий         113
Коричневый    107
Красный        63
Зелёный        34
Бежевый        24
Голубой        14
Оранжевый      13
Бордовый       13
Жёлтый         11
Золотой         7
Фиолетовый      6
Пурпурный       6
Name: count, dtype: int64

объединим мелкие категории цветов (меньше 15)

In [59]:
color_mapping = {
    'Чёрный': 'Чёрный',
    'Белый': 'Белый',
    'Серый': 'Серый',
    'Серебряный': 'Серебряный',
    'Синий': 'Синий',
    'Коричневый': 'Коричневый',
    'Красный': 'Красный',
    'Зелёный': 'Зелёный',
    'Бежевый': 'Бежевый',
    'Голубой': 'Другой',
    'Оранжевый': 'Другой',
    'Бордовый': 'Другой',
    'Жёлтый': 'Другой',
    'Золотой': 'Другой',
    'Фиолетовый': 'Другой',
    'Пурпурный': 'Другой'
}


df['color'] = df['color'].map(color_mapping).fillna('Другой')

#### Руль

In [60]:
df['wheel'].value_counts(dropna=False)

wheel
Левый     1676
Правый      22
Name: count, dtype: int64

все хорошо с данной колонкой

#### VIN или номер кузова

In [61]:
df['number'].value_counts(dropna=False)

number
XW8Z*************    54
Z94C*************    50
XWEG*************    46
WVWZ*************    45
XWEP*************    45
                     ..
TMBL*************     1
WF0E*************     1
WBAY*************     1
5UXZ*************     1
LS4A*************     1
Name: count, Length: 452, dtype: int64

Номер кузова сам по себе не несет логический смысл, поётому можем удалить данную колонку

In [62]:
df.drop(columns=['number'], inplace=True)

#### Обмен

In [63]:
df['exchange'].value_counts(dropna=False)

exchange
Не интересует    684
Возможен         636
0                378
Name: count, dtype: int64

переименуем значение 0 в другое

In [64]:
df['exchange'] = df['exchange'].map({'Не интересует': 'Не интересует', '0': 'другое', 'Возможен':'Возможен'})


#### Оценка

In [65]:
df['mark'].value_counts(dropna=False)

mark
NaN                                         209
899 000 ₽ — цена ниже Авито Оценки           17
1 099 000 ₽ — цена ниже Авито Оценки         12
1 199 000 ₽ — цена ниже Авито Оценки         12
799 000 ₽ — цена ниже Авито Оценки           11
                                           ... 
1 165 000 ₽ — цена выше Авито Оценки          1
195 000 ₽ — цена выше Авито Оценки            1
2 840 000 ₽ — цена выше Авито Оценки          1
1 030 000 ₽ — соответствует оценке Авито      1
1 155 000 ₽ — соответствует оценке Авито      1
Name: count, Length: 850, dtype: int64


из данной колонки мы хотим получить значение ниже / выше / соответствует оценки

In [66]:
df['mark'].fillna('нет оценки', inplace=True)


def categorize_mark(value):
    if 'ниже' in value:
        return 'ниже'
    elif 'выше' in value:
        return 'выше'
    elif 'соответствует' in value:
        return 'соответствует'
    else:
        return value  

df['mark'] = df['mark'].apply(categorize_mark)

#### ДТП

In [67]:
df['accident'].value_counts(dropna=False)

accident
ДТП не найдены    910
NaN               788
Name: count, dtype: int64

Данную колонку сделаем бинарной, где если ДТП не найдены будет значение 1, Nan – 0

In [68]:
df['accident'].fillna(0, inplace=True)
df['accident'] = df['accident'].map({'ДТП не найдены':1, 0:0})

#### другое_история

In [69]:
df['other_history'].value_counts(dropna=False)

other_history
NaN                     525
1 владелец по ПТС       471
2 владельца по ПТС      291
3 владельца по ПТС      177
4 владельца по ПТС       74
5 владельцев по ПТС      59
6 владельцев по ПТС      37
7 владельцев по ПТС      20
8 владельцев по ПТС      18
9 владельцев по ПТС      10
10 владельцев по ПТС      9
13 владельцев по ПТС      3
12 владельцев по ПТС      1
14 владельцев по ПТС      1
15 владельцев по ПТС      1
11 владельцев по ПТС      1
Name: count, dtype: int64

Выберем отсюда число владельцев


#### тех_история

In [70]:
df['tech_history'].value_counts(dropna=False)

tech_history
NaN                             644
1 запись с описанием работ      203
2 записи с описанием работ      126
3 записи с описанием работ       84
4 записи с описанием работ       76
5 записей с описанием работ      63
8 записей с описанием работ      54
6 записей с описанием работ      53
7 записей с описанием работ      49
9 записей с описанием работ      47
11 записей с описанием работ     33
12 записей с описанием работ     30
10 записей с описанием работ     28
13 записей с описанием работ     24
14 записей с описанием работ     21
17 записей с описанием работ     15
18 записей с описанием работ     15
15 записей с описанием работ     14
16 записей с описанием работ     13
19 записей с описанием работ     10
28 записей с описанием работ      8
27 записей с описанием работ      8
20 записей с описанием работ      8
21 запись с описанием работ       7
25 записей с описанием работ      6
22 записи с описанием работ       6
26 записей с описанием работ      5
30 записей с оп

Выберем число записей с описанием работ

#### sold

In [71]:
df['sold'].value_counts(dropna=False)

sold
NaN                           596
1 раз продавался на Авито     431
2 раза продавался на Авито    281
3 раза продавался на Авито    156
4 раза продавался на Авито    108
5 раз продавался на Авито      61
6 раз продавался на Авито      24
7 раз продавался на Авито      20
8 раз продавался на Авито      13
10 раз продавался на Авито      3
12 раз продавался на Авито      2
11 раз продавался на Авито      2
9 раз продавался на Авито       1
Name: count, dtype: int64

Можем заметить, что в признаках sold, тех история и лругое история одинаковый тип обработки – нам надо взять число и заполнить nan медианой (так как среднее может давать смещение из-за большого числа пустых значений). Также вспомним, что мы уже писали такую функцию для объёема двигателя (get_volume), можем ее использовать снова.

In [72]:
df['sold'] = df['sold'].apply(get_volume)
df['other_history'] = df['other_history'].apply(get_volume)
df['tech_history'] = df['tech_history'].apply(get_volume)

imputer = SimpleImputer(strategy='median')


columns_to_impute = ['sold', 'other_history', 'tech_history']

df[columns_to_impute] = imputer.fit_transform(df[columns_to_impute])


df[columns_to_impute] = df[columns_to_impute].astype(int)


/usr/local/anaconda3/lib/python3.10/site-packages/sklearn/utils/validation.py:767: FutureWarning:

is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.

/usr/local/anaconda3/lib/python3.10/site-packages/sklearn/utils/validation.py:605: FutureWarning:

is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.

/usr/local/anaconda3/lib/python3.10/site-packages/sklearn/utils/validation.py:614: FutureWarning:

is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.

/usr/local/anaconda3/lib/python3.10/site-packages/sklearn/utils/validation.py:767: FutureWarning:

is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.

/usr/local/anaconda3/lib/python3.10/site-packages/sklearn/utils/validation.py:605: FutureWarning:

is_sparse is deprecated and will be r

#### Метро

In [73]:
df['subway'].value_counts(dropna=False)

subway
Аннино:6–10 мин.;Лесопарковая:11–15 мин.;Улица Старокачаловская:21–30 мин.    73
Царицыно:от 31 мин.;Пражская:от 31 мин.;Кантемировская:от 31 мин.             65
Южная:11–15 мин.;Чертановская:21–30 мин.;Пражская:21–30 мин.                  62
Улица академика Янгеля:до 5 мин.;Аннино:11–15 мин.;Лесопарковая:21–30 мин.    62
Варшавская:до 5 мин.;Каховская:16–20 мин.;Чертановская:21–30 мин.             60
                                                                              ..
Семёновская:21–30 мин.;Шоссе Энтузиастов:21–30 мин.;Измайлово:21–30 мин.       1
Озёрная:16–20 мин.;Юго-Западная:16–20 мин.;Мичуринский проспект:21–30 мин.     1
Саларьево:11–15 мин.;Филатов луг:от 31 мин.;Румянцево:от 31 мин.               1
Марксистская:11–15 мин.;Таганская:11–15 мин.;Площадь Ильича:11–15 мин.         1
Автозаводская:до 5 мин.;ЗИЛ:16–20 мин.;Дубровка:21–30 мин.                     1
Name: count, Length: 519, dtype: int64

#### price

In [74]:
df['price'].value_counts(dropna=False)

price
899 000 ₽      24
1 099 000 ₽    24
1 199 000 ₽    21
999 000 ₽      20
1 295 000 ₽    20
               ..
5 250 000 ₽     1
3 680 000 ₽     1
1 648 000 ₽     1
2 100 000 ₽     1
1 155 000 ₽     1
Name: count, Length: 623, dtype: int64

Надо привести в числовой формат и убрать знак рубля в конце

In [75]:
df['price'] = df['price'].apply(lambda row: int(row.replace(' ', '').replace('₽', '')))

'Complete set',
       'Body Type', 'Color', 'Steering Wheel', 'VIN or Body number', 'Exchange', 'Rating',
       'Accident', 'other_history', 'tech_history', 'sold', 'Metro', 'pts_type',
       'price'

In [76]:
df.head(5)

title  year            age     run  history  \
0               Kia K900 3.3 AT  2019             II   91037      8.0   
1  Mercedes-Benz V-класс 2.1 AT  2017           W447   95250     19.0   
2         Opel Astra GTC 1.4 AT  2013   J рестайлинг  113000      6.0   
3      Volkswagen Tiguan 2.0 AT  2015   I рестайлинг  165000      9.0   
4              Ford Kuga 1.5 AT  2018  II рестайлинг  103000     16.0   

  owners_pts              modification  volume engine_type transmission  ...  \
0          2                3.3 4WD AT     3.3      Бензин      Автомат  ...   
1          2  V 250 d 2.1 4MATIC AT L3     2.1      Дизель      Автомат  ...   
2          2              1.4 Turbo AT     1.4      Бензин      Автомат  ...   
3          2        2.0 TDI 4Motion AT     2.0      Дизель      Автомат  ...   
4          1       1.5 EcoBoost 4WD AT     1.5      Бензин      Автомат  ...   

        exchange           mark accident other_history tech_history sold  \
0       Возможен  соответствует        1             2            4    1   
1  Не интересует  соответствует        0             2           19    2   
2       Возможен           ниже        1             2            5    3   
3       Возможен           ниже        0             2            5    1   
4         другое  соответствует        0             1           10    1   

                                              subway  pts_type    price  \
0  Аннино:6–10 мин.;Лесопарковая:11–15 мин.;Улица...    Другое  2999000   
1  Проспект Вернадского:до 5 мин.;Новаторская:16–...  Оригинал  5650000   
2  Южная:11–15 мин.;Чертановская:21–30 мин.;Пражс...    Другое   699000   
3  Волоколамская:16–20 мин.;Мякинино:от 31 мин.;Т...    Другое  1195000   
4  Ховрино:до 5 мин.;Беломорская:21–30 мин.;Речно...  Оригинал  1780000   

   horse_power  
0          249  
1          190  
2          140  
3          140  
4          150  

[5 rows x 24 columns]

In [77]:
df.isna().sum()

title             0
year              0
age               0
run               0
history           0
owners_pts        0
modification      0
volume            0
engine_type       0
transmission      0
gear              0
body              0
color             0
wheel             0
exchange          0
mark              0
accident          0
other_history     0
tech_history      0
sold              0
subway           23
pts_type          0
price             0
horse_power       0
dtype: int64

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1698 entries, 0 to 1803
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   title          1698 non-null   object 
 1   year           1698 non-null   int64  
 2   age            1698 non-null   object 
 3   run            1698 non-null   int64  
 4   history        1698 non-null   float64
 5   owners_pts     1698 non-null   object 
 6   modification   1698 non-null   object 
 7   volume         1698 non-null   float64
 8   engine_type    1698 non-null   object 
 9   transmission   1698 non-null   object 
 10  gear           1698 non-null   object 
 11  body           1698 non-null   object 
 12  color          1698 non-null   object 
 13  wheel          1698 non-null   object 
 14  exchange       1698 non-null   object 
 15  mark           1698 non-null   object 
 16  accident       1698 non-null   int64  
 17  other_history  1698 non-null   int64  
 18  tech_history 

### Сохранение результатов

Итого получилось 1698 строк и 22 признака. Сбросим индексы и сохраним результат

In [79]:
df.reset_index(drop=True, inplace=True)

In [80]:
df.to_csv('datasets/clear_dataset.csv')